### ISSUE TITLE
DacPackage.UpdateModel(TSqlModel) corrupts package: 'Origin.xml' is missing or corrupted.

### ISSUE THREAD

**Issue by smichtch** [microsoft/DacFx#242]
- SqlPackage or DacFx Version: Microsoft.SqlServer.DacFx 161.8089.0
- .NET Framework (Windows-only) or .NET Core: .NET 6.0
- Environment (local platform and source/target platforms):

**Steps to Reproduce:**
```csharp
// NOTE: The original .dacpac is generated by a .sqlproj - this is key!
var dacpacFromSqlproj = "OutputOfSqlproj.dacpac";

using var buffer = new MemoryStream();
using (var temp = File.OpenRead(dacpacFromSqlproj))
{
    temp.CopyTo(buffer); // to avoid modifying the original .dacpac file
}

buffer.Position = 0;
using (var dacpac = DacPackage.Load(buffer, DacSchemaModelStorageType.Memory, FileAccess.ReadWrite))
{
    dacpac.UpdateModel(
        new TSqlModel(SqlServerVersion.SqlAzure, new TSqlModelOptions()),
        new PackageMetadata());
}

buffer.Position = 0;
// DacException: 'Origin.xml' is missing or corrupted. ---> System.Xml.XmlException: Data at the root level is invalid. Line 23, position 13.
DacPackage.Load(buffer);
```

```
Microsoft.SqlServer.Dac.DacServicesException: Could not load package. ---> Microsoft.Data.Tools.Schema.Sql.Dac.DacException: 'Origin.xml' is missing or corrupted. ---> System.Xml.XmlException: Data at the root level is invalid. Line 23, position 13.

XmlTextReaderImpl.Throw(Exception e)
XmlTextReaderImpl.Throw(String res, String arg)
XmlTextReaderImpl.ParseRootLevelWhitespace()
XmlTextReaderImpl.ParseDocumentContent()
XmlTextReaderImpl.Read()
XContainer.ReadContentFrom(XmlReader r)
XContainer.ReadContentFrom(XmlReader r, LoadOptions o)
XDocument.Load(XmlReader reader, LoadOptions options)
XmlUtilities.LoadWithProhibitDtd(Stream stream, LoadOptions options)
DacOrigin.Deserialize(Stream stream)
--- End of inner exception stack trace ---
DacOrigin.Deserialize(Stream stream)
DacOrigin.Open(Stream stream)
DacUtilities.GetDacOrigin(SqlPackage package)
DacPackage.ContainsExportedData(IPackageSource packageSource)
DacPackage.ctor(IPackageSource packageSource, FileAccess packageAccess)
DacPackage.Load(Stream stream, DacSchemaModelStorageType modelStorageType, FileAccess packageAccess)
--- End of inner exception stack trace ---
DacPackage.Load(Stream stream, DacSchemaModelStorageType modelStorageType, FileAccess packageAccess)
DacPackage.Load(Stream stream, DacSchemaModelStorageType modelStorageType)
DacPackage.Load(Stream stream)
```

This is Origin.xml from the initial .dacpac which was generated from a .sqlproj:
```xml
<?xml version="1.0" encoding="utf-8"?>
<DacOrigin xmlns="http://schemas.microsoft.com/sqlserver/dac/Serialization/2012/02">
  <PackageProperties>
    <Version>3.0.0.0</Version>
    <ContainsExportedData>false</ContainsExportedData>
    <StreamVersions>
      <Version StreamName="Data">2.0.0.0</Version>
      <Version StreamName="DeploymentContributors">1.0.0.0</Version>
    </StreamVersions>
  </PackageProperties>
  <Operation>
    <Identity>a480dbf8-d356-4c4c-b3f8-821257ea2806</Identity>
    <Start>2023-03-13T22:36:23.9372197-07:00</Start>
    <End>2023-03-13T22:36:26.0148593-07:00</End>
    <ProductName>Microsoft.Data.Tools.Schema.Tasks.Sql, Version=16.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a</ProductName>
    <ProductVersion>16.0.6389.0</ProductVersion>
    <ProductSchema>http://schemas.microsoft.com/sqlserver/dac/Serialization/2012/02</ProductSchema>
  </Operation>
  <Checksums>
    <Checksum Uri="/model.xml">707F09FB16524772A2013C8D93B7DF6D7A21CED27C4ECF4522999F07A4625427</Checksum>
  </Checksums>
  <ModelSchemaVersion>2.4</ModelSchemaVersion>
</DacOrigin>
```

This is Origin.xml in the MemoryStream after calling DacPackage.UpdateModel():
```xml
<?xml version="1.0" encoding="utf-8"?>
<DacOrigin xmlns="http://schemas.microsoft.com/sqlserver/dac/Serialization/2012/02">
  <PackageProperties>
    <Version>3.1.0.0</Version>
    <ContainsExportedData>false</ContainsExportedData>
    <StreamVersions>
      <Version StreamName="Data">2.0.0.0</Version>
      <Version StreamName="DeploymentContributors">1.0.0.0</Version>
    </StreamVersions>
  </PackageProperties>
  <Operation>
    <Identity>a480dbf8-d356-4c4c-b3f8-821257ea2806</Identity>
    <Start>2023-03-13T22:59:31.2164864-07:00</Start>
    <End>2023-03-13T22:59:31.2178705-07:00</End>
    <ProductName>Microsoft.SqlServer.Dac.Extensions, Version=16.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a</ProductName>
    <ProductVersion>16.1.8430.0</ProductVersion>
    <ProductSchema>http://schemas.microsoft.com/sqlserver/dac/Serialization/2012/02</ProductSchema>
  </Operation>
  <Checksums>
    <Checksum Uri="/model.xml">823FEC3213691E0DFA47674B49448153E9AF8AA7C1BE0ED858388111FC321213</Checksum>
  </Checksums>
  <ModelSchemaVersion>2.4</ModelSchemaVersion>
</DacOrigin>in>
```
Both Origin.xml files are 1114 chars long, note the 3 junk characters on the end, also note that `ProductName` changed from Microsoft.Data.Tools.Schema.Tasks.Sql (37 chars) to Microsoft.SqlServer.Dac.Extensions (34 chars).
Seems like it's replacing the file contents but neglecting to update the stream/buffer length.

**Did this occur in prior versions?  If not - which version(s) did it work in?**
Tried all versions 161.8089.0 (latest stable) through 161.8430.0-preview (latest preview), not sure when this was introduced.
(DacFx/SqlPackage/SSMS/Azure Data Studio)

**Comment by llali** [comment]
@smichtch thanks for reporting this. I know you mentioned you were not sure when the bug was introduced, but can you please confirm if this were ever worked for you before? And also, is this working when using .Net 4.6.2? Thanks

**Comment by smichtch** [comment]
Ok so on net472, net48 and net481 (don't have net462 on hand to confirm) the last `DacPackage.Load(buffer);` call succeeds (I haven't checked the file package contents).
The bug repros when targeting net6.0, net7.0 and netcoreapp3.1.
Tried with Microsoft.SqlServer.DacFx 161.8089.0, 161.8430.0-preview, same results with both versions.

It's the first time I tried using the UpdateModel() API so I can't say it worked before.

**Comment by smichtch** [comment]
I just tried again with DacFx 162.0.34-preview, the original error appears to be fixed but still can't use the dacpac after updating the T-SQL model, updated repro code:

    // NOTE: The original .dacpac is generated by a .sqlproj - this is key!
    var dacpacFromSqlproj = Path.Combine(TestContext.TestDeploymentDir, "OutputOfSqlproj.dacpac");

    using var buffer = new MemoryStream();
    using (var temp = File.OpenRead(dacpacFromSqlproj))
    {
        temp.CopyTo(buffer); // to avoid modifying the original .dacpac file
    }

    buffer.Position = 0;
    using (var dacpac1 = DacPackage.Load(buffer, DacSchemaModelStorageType.Memory, FileAccess.ReadWrite))
    {
        dacpac1.UpdateModel(
            new TSqlModel(SqlServerVersion.SqlAzure, new TSqlModelOptions()),
            new PackageMetadata());
    }

    buffer.Position = 0;
    using var dacpac2 = DacPackage.Load(buffer);

    var dacServices = new DacServices(@"Data Source=(LocalDB)\MSSQLLocalDB;Integrated Security=true");

    // Next line throws:
    // Microsoft.SqlServer.Dac.DacServicesException: An error occurred during deployment plan generation.Deployment cannot continue.
    // ---> Microsoft.Data.Tools.Schema.Sql.Deployment.DeploymentFailedException: The calculated checksum for model.xml in the package is different from the saved checksum.
    // ---> Microsoft.Data.Tools.Schema.Sql.Build.SqlPackageException: The calculated checksum for model.xml in the package is different from the saved checksum.
    dacServices.GenerateDeployScript(dacpac2, "TargetDatabaseName");

**Comment by smichtch** [comment]
@llali, @zijchen should I open a new issue for this new error? Seems related to the first one.

**Comment by zijchen** [comment]
DacOrigin has a checksum of the model.xml inside the dacpac. Normally we only see this error when a schema model is manually modified, then the checksum needs to be re-calculated and origin.xml updated. Can you tell us more about the scenario you're trying to accomplish with the DacPackage?

**Comment by smichtch** [comment]
@zijchen, we want to modify the model to inject `CREATE USER`/`ALTER ROLE ADD MEMBER` statements which vary depending on the environment (and .sqlproj doesn't allow SQLCMD variables in those statements). I'm using an empty `TSqlModel` here for simplicity/min-repro, but shouldn't calling `DacPackage.UpdateModel()` result in a usable package?

**Comment by zijchen** [comment]
@smichtch I can reproduce this. Interestingly `DacPackage.UpdateModel()` actually tries to recalculate the checksum, but `DacServices` still complains about it when trying to publish/generate script. I will investigate further.

**Comment by zijchen** [comment]
@smichtch Could you create a new issue the checksum mismatch? So far it seems the checksum from UpdateModel is different than the one that's calculated when the DacPackage is reloaded.

**Comment by Ericvf** [comment]
We've noticed this issue a long time ago. Not sure exactly where the issue lies, we have tried to even fix it in this package. The Origin.xml does get corrupted, there is a bug somewhere probably in Dotnet. The solution for us has been to compile the project as net48. Unfortunately we can longer support that and therefore we are looking into this again. 

Curiously we have other dacpacs which work fine. This only breaks in certain circumstances (but is repeatable).

**Comment by Ericvf** [comment]
![image](https://github.com/microsoft/DacFx/assets/7957203/17d38600-8c08-4f9f-92c6-a2a4faccdd27)


Here you can see an example of the broken XML in Origin.xml (look at the closing end-tag)

**Comment by Ericvf** [comment]
BTW. I'm now using the following workaround:

```csharp
// OLD CODE
using (var dacPackage = DacPackage.Load(dacpacPath, DacSchemaModelStorageType.Memory, FileAccess.ReadWrite))
{
    logger.LogInformation("Updating model and writing to file...");
    DacPackageExtensions.UpdateModel(dacPackage, model, new PackageMetadata
    {
        Name = dacpacPath,
    });
}
```

```csharp
// NEW CODE
using (var dacPackage = DacPackage.Load(dacpacPath, DacSchemaModelStorageType.File, FileAccess.Read))
{
    logger.LogInformation("Loading external package and extracting pre/post deployment scripts");

    if (dacPackage.PreDeploymentScript != null)
    {
        preDeploymentScripts = new StreamReader(dacPackage.PreDeploymentScript).ReadToEnd();
    }

    if (dacPackage.PostDeploymentScript != null)
    {
        postDeploymentScripts = new StreamReader(dacPackage.PostDeploymentScript).ReadToEnd();
    }
}

logger.LogInformation("Creating new deployment package");
DacPackageExtensions.BuildPackage(dacpacPath, model, new PackageMetadata
{
    Name = dacpacPath,
});

logger.LogInformation("Adding pre/post deployment scripts");
AddPrePostDeploymentScripts(dacpacPath, preDeploymentScripts, postDeploymentScripts);

// ... 
private void AddPrePostDeploymentScripts(string dacpacPath, string preDeploymentScripts, string postDeploymentScripts)
{
    using (var package = Package.Open(dacpacPath, FileMode.Open, FileAccess.ReadWrite))
    {
        if (!string.IsNullOrEmpty(preDeploymentScripts))
        {
            var part = package.CreatePart(new Uri("/predeploy.sql", UriKind.Relative), "text/plain");

            using (var stream = part.GetStream())
            {
                stream.Write(Encoding.UTF8.GetBytes(preDeploymentScripts), 0, preDeploymentScripts.Length);
            }
        }

        if (!string.IsNullOrEmpty(postDeploymentScripts))
        {
            var part = package.CreatePart(new Uri("/postdeploy.sql", UriKind.Relative), "text/plain");

            using (var stream = part.GetStream())
            {
                stream.Write(Encoding.UTF8.GetBytes(postDeploymentScripts), 0, postDeploymentScripts.Length);
            }
        }

        package.Close();
    }
}
```

**Comment by smichtch** [comment]
> @smichtch Could you create a new issue the checksum mismatch? So far it seems the checksum from UpdateModel is different than the one that's calculated when the DacPackage is reloaded.
Done, filed #281.

> We've noticed this issue a long time ago. Not sure exactly where the issue lies, we have tried to even fix it in this package. The Origin.xml does get corrupted, there is a bug somewhere probably in Dotnet. The solution for us has been to compile the project as net48. Unfortunately we can longer support that and therefore we are looking into this again.
> 
> Curiously we have other dacpacs which work fine. This only breaks in certain circumstances (but is repeatable).
As I mentioned in the original post it only occurs when the initial .dacpac was generated using a .sqlproj and then modified using the DacFx library (the DacOrigin/Operation/ProductName string in origin.xml changes and becomes 3 chars shorter). I suspect Stream behavior changed between .NET Framework to .NET Core so that the length is not updated automatically when the new origin.xml content is written.
(The mismatching checksum may be related to non-deterministic GetHashCode() behavior in .NET Core versus .NET Core, but this is just speculation)

**Comment by Ericvf** [comment]
Yeah, this is also my feeling. I noticed that the Serialize method in DacOrigin does not specify the length to be written. It could be due to something like that.

**Comment by zijchen** [comment]
@Ericvf The original issue (with the extra characters breaking the XML) should be fixed in the latest release (162.0.52). I'm investigating the checksum issue right now.

I'll close this issue now, but feel free to reopen if you still see the broken XML bug. Please use #281 to discuss about the checksum issue.
